tools/runfiles/runfiles.rs

```diff
@@ -85,46 +95,78 @@ impl Runfiles {
     /// validity and that the path exists.
+    /// @deprecated - this is not bzlmod-aware. Prefer the `Rlocation!` macro or `rlocation_from`
     pub fn rlocation(&self, path: impl AsRef<Path>) -> PathBuf {
         let path = path.as_ref();
         if path.is_absolute() {
             return path.to_path_buf();
         }
-        match &self.mode {
-            Mode::DirectoryBased(runfiles_dir) => runfiles_dir.join(path),
-            Mode::ManifestBased(path_mapping) => path_mapping
-                .get(path)
-                .unwrap_or_else(|| {
-                    panic!("Path {} not found among runfiles.", path.to_string_lossy())
-                })
-                .clone(),
+        raw_rlocation(&self.mode, path)
+    }
+
+    /// Returns the runtime path of a runfile.
+    ///
+    /// Runfiles are data-dependencies of Bazel-built binaries and tests.
+    /// The returned path may not be valid. The caller should check the path's
+    /// validity and that the path exists.
+    ///
+    /// Typically this should be used via the `rlocation!` macro to properly set source_repo.
+    pub fn rlocation_from(&self, path: impl AsRef<Path>, source_repo: &str) -> PathBuf {
+        let path = path.as_ref();
+        if path.is_absolute() {
+            return path.to_path_buf();
+        }
+
+        let parts: Vec<&str> = path
+            .to_str()
+            .expect("Should be valid UTF8")
+            .splitn(2, '/')
+            .collect();
+        if parts.len() == 2 {
+            let key: (String, String) = (source_repo.into(), parts[0].into());
+            if let Some(target_repo_directory) = self.repo_mapping.get(&key) {
+                return raw_rlocation(
+                    &self.mode,
+                    target_repo_directory.to_owned() + "/" + parts[1],
+                );
+            };
         }
+        raw_rlocation(&self.mode, path)
     }
+}
 
-    /// Returns the canonical name of the caller's Bazel repository.
-    pub fn current_repository(&self) -> &str {
-        // This value must match the value of `_RULES_RUST_RUNFILES_WORKSPACE_NAME`
-        // which can be found in `@rules_rust//tools/runfiles/private:workspace_name.bzl`
-        env!("RULES_RUST_RUNFILES_WORKSPACE_NAME")
+fn raw_rlocation(mode: &Mode, path: impl AsRef<Path>) -> PathBuf {
+    let path = path.as_ref();
+    match mode {
+        Mode::DirectoryBased(runfiles_dir) => runfiles_dir.join(path),
+        Mode::ManifestBased(path_mapping) => path_mapping
+            .get(path)
+            .unwrap_or_else(|| panic!("Path {} not found among runfiles.", path.to_string_lossy()))
+            .clone(),
     }
 }
 
+fn parse_repo_mapping(path: PathBuf) -> io::Result<RepoMapping> {
+    Ok(std::fs::read_to_string(path)?
+        .lines()
+        .map(|line| {
+            let parts: Vec<String> = line.splitn(3, ',').map(String::from).collect();
+            ((parts[0].clone(), parts[1].clone()), parts[2].clone())
```

Comment:
It would be good to error rather than panic if this file is malformed and doesn't contain 3 elements

Reply:
done, I also fixed the duplicative string copying while I was here